monkey/infection_monkey/master/ip_scanner.py

```diff
@@ -62,9 +67,17 @@ def _scan_addresses(
         try:
             while not stop.is_set():
                 address = addresses.get_nowait()
-                logger.info(f"Scanning {address.ip}")
+                logger.info(f"Scanning {address.ip} {type(address)}")
 
                 ping_scan_data = self._puppet.ping(address.ip, icmp_timeout)
+                self._agent_event_queue.publish(
+                    PingScanEvent(
+                        source=get_agent_id(),
+                        target=IPv4Address(address.ip),
+                        scan_data=ping_scan_data,
+                    )
+                )
```

Comment:
This is being done in the wrong place. We should clarify an architectural boundary:

_Where possible, event objects are generated and published by the components that generated them._

There is an architectural boundary between the master and puppet such that the master doesn't concern itself with the specific technical details of the puppet. The master commands the agent to "ping" a target. The specifics of how that happens are left to the puppet. In other words, **only the puppet is capable of properly creating events**. The master does not have the necessary information to set MITRE ATT&CK tags, or zero trust information. It also doesn't know how many packets were sent; maybe multiple events should be generated.

One specific issue with the way that this is implemented is that the timestamp of the event will be wrong. While all timestamps will be an approximation, we should make our best effort to get as close as possible. If a `ping()` is requested by the master at t=0, the master has no way of knowing when the ping will be sent. If, for example, the ping timeout is set to 1 second (the default, I believe) and the puppet sends a packet at t=1, if no response is received, the puppet will return at approximately t=2 and the current implementation would set the timestamp of the event to 2, but the correct value is 1.

In summary, the component that performs the `ping` operation should generate and send the events.

Reply:
This will couple the component logic to the data transfer logic. Timestamp and other relevant information can be added to `ping_scan_data` and then unpacked into events by the master. This would allow changing components/data transfer mechanism independently. But if we want to make components "standalone", then the easiest way to do that is to give components both responsibilities. ATM I don't see any infrastructure to do that. Passing down the event queue in parameters is not a good way, because you're tying the component to the puppet and we want users to be able to easily modify/create components. Creating a dependency injection based on type hints is not intuitive and an overkill. I think we should do something like flask does with `request`. Import a global `event_queue` and push events to it.

Reply:
> This will couple the component logic to the data transfer logic.

Maybe, depends how you define "data transfer".

> Timestamp and other relevant information can be added to ping_scan_data and then unpacked into events by the master.

What if multiple events are generated? What if the component generates multiple types of events? The master shouldn't and can't be aware of every type of event that a component could generate.

If there's a concern over coupling the event queue to low-level components, we could have the components return a `Sequence[AbstractAgentEvent]` and have the caller do the publishing.

> Passing down the event queue in parameters is not a good way, because you're tying the component to the puppet
No, you're tying the component to `IAgentEventQueue`. The component still has no knowlegde of the puppet.

> I think we should do something like flask does with response. Import a global event_queue and push events to it.
Certainly not. Nothing global! Flask has a different set of requirements and use cases than we do.

We can talk more and come to a consensus on team time.